File: js/frame-assembler.js
import { decodeFrame } from "./protocol.js";
import { START, END } from "./symbols.js";

/**
 * Estado inicial del ensamblador: todavia no vio el marcador de inicio.
 */
export function createAssemblyState() {
  return { receiving: false, buffer: [] };
}

/**
 * Avanza el ensamblador con un evento de simbolo confirmado (ver
 * js/matcher.js). El marcador START (re)arranca la acumulacion desde cero
 * -incluso si ya estaba recibiendo, asi una transmision repetida siempre
 * puede "ganar" sobre una anterior que quedo a mitad de camino-. Los bytes
 * de datos se ignoran si todavia no se vio el START. El marcador END cierra
 * la trama y la decodifica.
 * @param {{receiving: boolean, buffer: number[]}} state
 * @param {{value: number|string}} event
 * @returns {{state: object, done: null | {ok: true, text: string} | {ok: false, error: string}}}
 */
export function advanceAssembly(state, event) {
  const { value } = event;

  if (value === START) {
    return { state: { receiving: true, buffer: [] }, done: null };
  }

  if (!state.receiving) {
    return { state, done: null };
  }

  if (value === END) {
    const result = decodeFrame(state.buffer);
    return { state: createAssemblyState(), done: result };
  }

  return { state: { receiving: true, buffer: [...state.buffer, value] }, done: null };
}

File: android/app/src/main/java/com/memetransfer/receiver/protocol/FrameAssembler.kt
package com.memetransfer.receiver.protocol

/** Puerto directo de js/frame-assembler.js. */
data class AssemblyState(val receiving: Boolean = false, val buffer: List<Int> = emptyList())

fun createAssemblyState(): AssemblyState = AssemblyState()

/**
 * Avanza el ensamblador con un simbolo confirmado (ver SymbolStream). El
 * simbolo START (re)arranca la acumulacion desde cero -incluso si ya
 * estaba recibiendo, asi una transmision repetida siempre puede "ganar"
 * sobre una anterior que quedo a mitad de camino-. Los bytes de datos se
 * ignoran si todavia no se vio el START. El simbolo END cierra la trama y
 * la decodifica.
 * @return el nuevo estado, y el resultado de decodeFrame si END cerro una trama (o null si no).
 */
fun advanceAssembly(state: AssemblyState, value: Any): Pair<AssemblyState, DecodeResult?> {
    if (value == START) {
        return AssemblyState(receiving = true, buffer = emptyList()) to null
    }

    if (!state.receiving) {
        return state to null
    }

    if (value == END) {
        val result = decodeFrame(state.buffer)
        return createAssemblyState() to result
    }

    val byte = value as Int
    return AssemblyState(receiving = true, buffer = state.buffer + byte) to null
}
